crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -216,21 +249,41 @@ impl QueryLimitsChecker {
         while !que.is_empty() {
             // Signifies the start of a new level
             cost.depth += 1;
-            check_limits(limits, cost.num_nodes, cost.depth, None)?;
+            check_limits(limits, cost.num_nodes, cost.depth, cost.output_nodes, None)?;
             while level_len > 0 {
                 // Ok to unwrap since we checked for empty queue
                 // and level_len > 0
-                let curr_sel = que.pop_front().unwrap();
+                let ToVisit {
+                    selection: curr_sel,
+                    parent_node_count,
+                } = que.pop_front().unwrap();
 
                 match &curr_sel.node {
                     Selection::Field(f) => {
                         check_directives(&f.node.directives)?;
+
+                        let current_count =
+                            estimate_output_nodes_for_curr_node(f, variables, limits)
+                                * parent_node_count;
+
+                        cost.output_nodes += current_count;
+
                         for field_sel in f.node.selection_set.node.items.iter() {
-                            que.push_back(field_sel);
+                            que.push_back(ToVisit {
+                                selection: field_sel,
+                                parent_node_count: current_count,
```

Comment:
I think the parent count you use here should depend on which field you are recursing into.  Take the example of a connection query that also requests the page info.  Currently, whether you are recursing into `edges`, or `nodes` or `pageInfo` you are going to apply the multiplier from it being a connection, but this is not correct -- we are over- counting `pageInfo`.

We should only use the current count if we are recursing into `edges` or `nodes`, otherwise, we should keep using the existing parent count.

Reply:
Ah... right, I overlooked this. Thanks for catching. There are n nodes or edges but just 1 pageInfo for the connection.